src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1511 @@
+            referenceExpressionNode = false;
+            break;
+        }
+
+        // Scan node
+        super.scan(tree);
+
+        // Sanity check
+        Assert.check(checkInvariants(true, referenceExpressionNode));
+
+        // Discard any direct 'this' reference that can't be valid because its
+        // type is not compatible with the target class we're analyzing for.
+        // Note, tricky code could cast away and then cast back to evade this.
+        if (referenceExpressionNode) {
+
+            // We treat instance methods as having a "value" equal to their instance
+            Type type = tree.type;
+            Symbol sym = TreeInfo.symbolFor(tree);
+            if (sym != null &&
+                sym.kind == MTH &&
+                (sym.flags() & Flags.STATIC) == 0) {
+                type = sym.owner.type;
+            }
+
+            // If the expression type is incompatible with 'this', discard it
+            if (type != null && !targetClass.sym.isSubClass(type.tsym, types))
+                refs.remove(ExprRef.direct(depth));
+        }
+    }
+
+//
+// Visitor methods - Class Declarations
+//
+
+    @Override
+    public void visitClassDef(JCClassDecl tree) {
+        return;     // we're busy analyzing another class - skip
+    }
+
+//
+// Visitor methods - Variable Declarations
+//
+
+    @Override
+    public void visitVarDef(JCVariableDecl tree) {
+
+        // Skip if ignoring warnings for this field
+        if (suppressed.contains(tree.sym))
+            return;
+
+        // Scan initializer, if any
+        scan(tree.init);
+        if (isParamOrVar(tree.sym))
+            refs.replaceExprs(depth, direct -> new VarRef(tree.sym, direct));
+        else
+            refs.discardExprs(depth);           // we don't track fields yet
+    }
+
+//
+// Visitor methods - Methods
+//
+
+    @Override
+    public void visitMethodDef(JCMethodDecl tree) {
+        Assert.check(false);        // we should never get here
+    }
+
+    @Override
+    public void visitApply(JCMethodInvocation invoke) {
+
+        // Get method symbol
+        MethodSymbol sym = (MethodSymbol)TreeInfo.symbolFor(invoke.meth);
+
+        // Recurse on method expression
+        scan(invoke.meth);
+        boolean direct = refs.remove(ExprRef.direct(depth));
+        boolean indirect = refs.remove(ExprRef.indirect(depth));
+
+        // Determine if method receiver represents a possible reference
+        RefSet<ThisRef> receiverRefs = RefSet.newEmpty();
+        if (sym != null && !sym.isStatic()) {
+            if (direct)
+                receiverRefs.add(ThisRef.direct());
+            if (indirect)
+                receiverRefs.add(ThisRef.indirect());
+        }
+
+        // If "super()": ignore - we don't try to track into superclasses.
+        // However, we do need to "invoke" non-static initializers/blocks.
+        Name name = TreeInfo.name(invoke.meth);
+        if (name == names._super) {
+            scanInitializers();
```

Comment:
it seems like the code scan initializers every time it finds a super() invocation, I guess that this scanning could be done once per class

Reply:
Yes... I did it that way is so that it doesn't require any adaptation if/when JDK-8194743 ever gets implemented. And it keeps the code a little simpler in exchange for a little redundancy.

I'm happy to fix this if you think it is necessary though.

Reply:
I'm OK either way we can revisit this later either as part of this PR or in a future one. I let it to your consideration

Reply:
Sounds good - thanks.

Reply:
Ah. I just realized that we need to do it your way because of the following bug:

Suppose you have a constructor and a field with initializer that both leak, but you have `@SuppressWarnings("this-escape")` on the constructor.

Then the leak for the field will never be reported because we never get to it.

I'll fix.

Reply:
yep, right in that case that bug wouldn't be reported